src/core/ngx_inet.c

```diff
@@ -942,6 +949,307 @@ ngx_parse_inet6_url(ngx_pool_t *pool, ngx_url_t *u)
+            pos = ngx_strlchr(url->addrs[i].name.data, url->addrs[i].name.data 
+                              + url->addrs[i].name.len, ']');
+            if (pos == NULL) {
+                continue;
+            }        
+            q.data = url->addrs[i].name.data + 1;
+            q.len = pos - q.data;
+        }
+        
+        len = ngx_snprintf(buf, 69, "%V|%d%N", &q, 
+                           url->addrs[i].sockaddr->sa_family) - buf;
+        ngx_write_fd(fd, buf, len);
+    }
+
+    ngx_close_file(fd);   
+    
+    if (ngx_rename_file(tpathf.data, dpathf.data) == NGX_FILE_ERROR){
+        ngx_log_error(NGX_LOG_EMERG, ngx_cycle->log, ngx_errno,
+                      ngx_rename_file_n" from \"%s\" to \"%s\" failed",
+                      tpathf.data, dpathf.data);
+        return NGX_ERROR;
+    }    
+    
+    return NGX_OK;
+}
+
+
+static ngx_int_t
+ngx_resolve_using_local(ngx_pool_t *pool, ngx_url_t **u, ngx_str_t path) 
+{
+    u_char                              *buf, *p, *s, *q, *l;
+    size_t                               len;
+    ngx_uint_t                           i;
+    ngx_str_t                            pf;
+    ngx_fd_t                             fd;
+    ngx_uint_t                           size, ip, sin_family;
+    ngx_file_info_t                      info;
+    ngx_url_t                           *url;
+    struct sockaddr_in                  *sin;
+#if (NGX_HAVE_INET6)
+    struct sockaddr_in6                 *sin6;
+    u_char                              *v6;
+#endif    
+    
+    url = *u;
+    if (path.data && path.data[path.len - 1] != '/') {
+        p = ngx_pcalloc(pool, path.len + 1);   
+        if (p == NULL) {
+            return NGX_ERROR;
+        }
+        
+        path.len = ngx_sprintf(p, "%V/", &path) - p;
+        path.data = p;
+    }
+    
+    pf.len = path.len + url->host.len;
+    pf.data = ngx_pcalloc(pool, pf.len + 1);
+    if (pf.data == NULL) {
+        return NGX_ERROR;
+    }
+
+    ngx_snprintf(pf.data, pf.len, "%V%*s",
+                 &path, url->host.len, url->host.data);
+
+    if (ngx_conf_full_name((ngx_cycle_t *) ngx_cycle, &pf, 0) != NGX_OK) {
+        return NGX_ERROR;
+    }
+
+    fd = ngx_open_file(pf.data, NGX_FILE_RDONLY, NGX_FILE_OPEN, 0);
+    if (fd == NGX_INVALID_FILE) {
+        return NGX_ERROR;
+    }
+
+    if (ngx_fd_info(fd, &info) == -1) {
+        ngx_log_error(NGX_LOG_ERR, ngx_cycle->log, ngx_errno,
+                      ngx_fd_info_n " \"%V\" failed", &pf);
+        return NGX_ERROR;
+    }
+
+    /* fix: maybe disk full */
+    size = ngx_file_size(&info);
+    if (size == 0) {
+        ngx_log_error(NGX_LOG_ERR, ngx_cycle->log, 0,
+                      "dns : the cache \"%V\" is empty file", &pf);
+        return NGX_ERROR;
+    }
+
+    buf = ngx_palloc(pool, size);
+    if (buf == NULL) {
+        ngx_close_file(fd);
+        return NGX_ERROR;
+    }
+
+    if (ngx_read_fd(fd, buf, size) == -1) {
+        return NGX_ERROR;
```

Comment:
miss ngx_close_file

Reply:
Yeah, It's should be close the fd in some exceptional cases.  
I have updated the code, PLS Review. Thx